File: backend/app/models/nutrition.py
"""Nutrition and meal models."""
import enum
import uuid
from datetime import datetime
from typing import Optional, List

from sqlalchemy import Enum, String, Float, Integer, ForeignKey, DateTime, Text, func
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class FoodSource(str, enum.Enum):
    """Source of food item data."""
    MANUAL = "manual"  # Manually entered
    OPEN_FOOD_FACTS = "open_food_facts"  # From OFF API
    BARCODE = "barcode"  # Scanned barcode -> OFF
    VISION = "vision"  # AI vision estimate
    CHAT = "chat"  # Added via chat assistant


class MealType(str, enum.Enum):
    """Type of meal."""
    BREAKFAST = "breakfast"
    LUNCH = "lunch"
    DINNER = "dinner"
    SNACK = "snack"
    OTHER = "other"


class Meal(Base):
    """A meal containing one or more food items."""

    __tablename__ = "meals"

    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("user_profiles.id", ondelete="CASCADE"), index=True
    )

    # Meal info
    name: Mapped[str] = mapped_column(String(200))
    meal_type: Mapped[MealType] = mapped_column(Enum(MealType), default=MealType.OTHER)
    timestamp: Mapped[datetime] = mapped_column(DateTime(timezone=True), index=True)

    # Aggregated totals (denormalized for quick queries)
    total_calories: Mapped[int] = mapped_column(Integer, default=0)
    total_protein_g: Mapped[float] = mapped_column(Float, default=0.0)
    total_carbs_g: Mapped[float] = mapped_column(Float, default=0.0)
    total_fat_g: Mapped[float] = mapped_column(Float, default=0.0)
    total_fiber_g: Mapped[Optional[float]] = mapped_column(Float, nullable=True)

    # Optional
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    photo_url: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)

    # Sync
    is_synced: Mapped[bool] = mapped_column(default=True)
    local_id: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)

    # Relationships
    items: Mapped[List["FoodItem"]] = relationship(
        "FoodItem", back_populates="meal", cascade="all, delete-orphan"
    )

    def recalculate_totals(self):
        """Recalculate total macros from items."""
        self.total_calories = sum(item.calories for item in self.items)
        self.total_protein_g = sum(item.protein_g for item in self.items)
        self.total_carbs_g = sum(item.carbs_g for item in self.items)
        self.total_fat_g = sum(item.fat_g for item in self.items)
        self.total_fiber_g = sum(
            item.fiber_g for item in self.items if item.fiber_g
        ) or None


class FoodItem(Base):
    """Individual food item within a meal."""

    __tablename__ = "food_items"

    meal_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("meals.id", ondelete="CASCADE"), index=True
    )

    # Food info
    name: Mapped[str] = mapped_column(String(200))
    source: Mapped[FoodSource] = mapped_column(Enum(FoodSource))

    # Quantity
    grams: Mapped[float] = mapped_column(Float)
    serving_size: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    serving_unit: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    servings: Mapped[float] = mapped_column(Float, default=1.0)

    # Macros (for the amount consumed)
    calories: Mapped[int] = mapped_column(Integer)
    protein_g: Mapped[float] = mapped_column(Float)
    carbs_g: Mapped[float] = mapped_column(Float)
    fat_g: Mapped[float] = mapped_column(Float)
    fiber_g: Mapped[Optional[float]] = mapped_column(Float, nullable=True)

    # Additional nutrition (optional)
    sodium_mg: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    sugar_g: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    saturated_fat_g: Mapped[Optional[float]] = mapped_column(Float, nullable=True)

    # Barcode/OFF data
    barcode: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    off_product_id: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    nutri_score_grade: Mapped[Optional[str]] = mapped_column(String(5), nullable=True)

    # Vision AI metadata
    confidence: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    portion_description: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)

    # Relationships
    meal: Mapped["Meal"] = relationship("Meal", back_populates="items")


class SavedFood(Base):
    """Saved/favorited foods for quick access."""

    __tablename__ = "saved_foods"

    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("user_profiles.id", ondelete="CASCADE"), index=True
    )

    # Food info
    name: Mapped[str] = mapped_column(String(200))
    brand: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    source: Mapped[FoodSource] = mapped_column(Enum(FoodSource))

    # Per 100g values
    calories_per_100g: Mapped[int] = mapped_column(Integer)
    protein_per_100g: Mapped[float] = mapped_column(Float)
    carbs_per_100g: Mapped[float] = mapped_column(Float)
    fat_per_100g: Mapped[float] = mapped_column(Float)
    fiber_per_100g: Mapped[Optional[float]] = mapped_column(Float, nullable=True)

    # Default serving
    default_serving_g: Mapped[float] = mapped_column(Float, default=100.0)
    serving_unit: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)

    # Barcode
    barcode: Mapped[Optional[str]] = mapped_column(String(50), index=True, nullable=True)

    # Metadata
    times_used: Mapped[int] = mapped_column(Integer, default=0)
    last_used_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)

    # Raw OFF data for reference
    off_data: Mapped[Optional[dict]] = mapped_column(JSONB, nullable=True)

File: Logged/Logged/Models/Nutrition.swift
import Foundation
import SwiftData

/// Source of food item data
enum FoodSource: String, Codable {
    case manual
    case openFoodFacts = "open_food_facts"
    case barcode
    case vision
    case chat
}

/// Type of meal
enum MealType: String, Codable, CaseIterable {
    case breakfast
    case lunch
    case dinner
    case snack
    case other

    var displayName: String {
        rawValue.capitalized
    }

    var icon: String {
        switch self {
        case .breakfast: return "sunrise.fill"
        case .lunch: return "sun.max.fill"
        case .dinner: return "moon.fill"
        case .snack: return "carrot.fill"
        case .other: return "fork.knife"
        }
    }
}

/// A meal containing one or more food items
@Model
final class Meal {
    @Attribute(.unique) var id: UUID
    var userId: UUID
    var localId: String?

    // Meal info
    var name: String
    var mealType: MealType
    var timestamp: Date
    var notes: String?
    var photoURL: String?

    // Aggregated totals
    var totalCalories: Int
    var totalProteinG: Double
    var totalCarbsG: Double
    var totalFatG: Double
    var totalFiberG: Double?

    // Relationships
    @Relationship(deleteRule: .cascade, inverse: \FoodItem.meal)
    var items: [FoodItem]

    // Sync
    var isSynced: Bool
    var createdAt: Date
    var updatedAt: Date

    init(
        id: UUID = UUID(),
        userId: UUID,
        name: String,
        mealType: MealType = .other,
        timestamp: Date = Date(),
        notes: String? = nil
    ) {
        self.id = id
        self.userId = userId
        self.localId = UUID().uuidString
        self.name = name
        self.mealType = mealType
        self.timestamp = timestamp
        self.notes = notes
        self.totalCalories = 0
        self.totalProteinG = 0
        self.totalCarbsG = 0
        self.totalFatG = 0
        self.items = []
        self.isSynced = false
        self.createdAt = Date()
        self.updatedAt = Date()
    }

    func recalculateTotals() {
        totalCalories = items.reduce(0) { $0 + $1.calories }
        totalProteinG = items.reduce(0) { $0 + $1.proteinG }
        totalCarbsG = items.reduce(0) { $0 + $1.carbsG }
        totalFatG = items.reduce(0) { $0 + $1.fatG }
        totalFiberG = items.compactMap { $0.fiberG }.reduce(0, +)
        updatedAt = Date()
    }

    func addItem(_ item: FoodItem) {
        items.append(item)
        recalculateTotals()
    }

    func removeItem(_ item: FoodItem) {
        items.removeAll { $0.id == item.id }
        recalculateTotals()
    }
}

/// Individual food item within a meal
@Model
final class FoodItem {
    @Attribute(.unique) var id: UUID
    var meal: Meal?

    // Food info
    var name: String
    var source: FoodSource

    // Quantity
    var grams: Double
    var servingSize: Double?
    var servingUnit: String?
    var servings: Double

    // Macros (for the amount consumed)
    var calories: Int
    var proteinG: Double
    var carbsG: Double
    var fatG: Double
    var fiberG: Double?

    // Additional nutrition
    var sodiumMg: Double?
    var sugarG: Double?
    var saturatedFatG: Double?

    // Barcode/OFF data
    var barcode: String?
    var offProductId: String?
    var nutriScoreGrade: String?

    // Vision AI metadata
    var confidence: Double?
    var portionDescription: String?

    var createdAt: Date

    init(
        id: UUID = UUID(),
        name: String,
        source: FoodSource = .manual,
        grams: Double,
        calories: Int,
        proteinG: Double,
        carbsG: Double,
        fatG: Double,
        fiberG: Double? = nil,
        servingSize: Double? = nil,
        servingUnit: String? = nil,
        servings: Double = 1.0,
        barcode: String? = nil,
        nutriScoreGrade: String? = nil,
        confidence: Double? = nil,
        portionDescription: String? = nil
    ) {
        self.id = id
        self.name = name
        self.source = source
        self.grams = grams
        self.calories = calories
        self.proteinG = proteinG
        self.carbsG = carbsG
        self.fatG = fatG
        self.fiberG = fiberG
        self.servingSize = servingSize
        self.servingUnit = servingUnit
        self.servings = servings
        self.barcode = barcode
        self.nutriScoreGrade = nutriScoreGrade
        self.confidence = confidence
        self.portionDescription = portionDescription
        self.createdAt = Date()
    }
}

/// Saved foods for quick access
@Model
final class SavedFood {
    @Attribute(.unique) var id: UUID
    var userId: UUID

    // Food info
    var name: String
    var brand: String?
    var source: FoodSource

    // Per 100g values
    var caloriesPer100g: Int
    var proteinPer100g: Double
    var carbsPer100g: Double
    var fatPer100g: Double
    var fiberPer100g: Double?

    // Default serving
    var defaultServingG: Double
    var servingUnit: String?

    // Barcode
    var barcode: String?

    // Usage tracking
    var timesUsed: Int
    var lastUsedAt: Date?

    var createdAt: Date

    init(
        id: UUID = UUID(),
        userId: UUID,
        name: String,
        brand: String? = nil,
        source: FoodSource = .manual,
        caloriesPer100g: Int,
        proteinPer100g: Double,
        carbsPer100g: Double,
        fatPer100g: Double,
        fiberPer100g: Double? = nil,
        defaultServingG: Double = 100,
        servingUnit: String? = nil,
        barcode: String? = nil
    ) {
        self.id = id
        self.userId = userId
        self.name = name
        self.brand = brand
        self.source = source
        self.caloriesPer100g = caloriesPer100g
        self.proteinPer100g = proteinPer100g
        self.carbsPer100g = carbsPer100g
        self.fatPer100g = fatPer100g
        self.fiberPer100g = fiberPer100g
        self.defaultServingG = defaultServingG
        self.servingUnit = servingUnit
        self.barcode = barcode
        self.timesUsed = 0
        self.createdAt = Date()
    }

    func toFoodItem(grams: Double) -> FoodItem {
        let multiplier = grams / 100.0
        return FoodItem(
            name: brand != nil ? "\(brand!) \(name)" : name,
            source: source,
            grams: grams,
            calories: Int(Double(caloriesPer100g) * multiplier),
            proteinG: proteinPer100g * multiplier,
            carbsG: carbsPer100g * multiplier,
            fatG: fatPer100g * multiplier,
            fiberG: fiberPer100g.map { $0 * multiplier },
            servingSize: defaultServingG,
            servingUnit: servingUnit,
            barcode: barcode
        )
    }
}
